File: src/main/GO/leetcode/lists/offer/Offer34.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/4 14:53
 */

func pathSum(root *TreeNode, target int) [][]int {
	if root == nil {
		return [][]int{}
	}
	list := make([][]int, 0)
	var dfs func(root *TreeNode, target int, temp []int)
	dfs = func(root *TreeNode, target int, temp []int) {
		if root == nil {
			return
		}
		if target-root.Val == 0 && root.Right == nil && root.Left == nil {
			temp = append(temp, root.Val)
			list = append(list, append([]int(nil), temp...))
			return
		} else {
			dfs(root.Right, target-root.Val, append(temp, root.Val))
			dfs(root.Left, target-root.Val, append(temp, root.Val))
		}
	}
	dfs(root, target, make([]int, 0))
	return list
}

File: java/src/leetcode/lists/offer/Offer34.java
package leetcode.lists.offer;

import common.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/12 23:27
 */
public class Offer34 {

    List<List<Integer>> res;

    public List<List<Integer>> pathSum(TreeNode root, int target) {
        res = new ArrayList<>();
        dfs(root, target, new ArrayList<>());
        return res;
    }

    private void dfs(TreeNode root, int target, List<Integer> temp) {
        if (root == null) {
            return;
        }
        if (target == root.val && root.left == null && root.right == null) {
            temp.add(root.val);
            res.add(new ArrayList<>(temp));
            temp.remove(temp.size() - 1);
            return;
        }
        temp.add(root.val);
        dfs(root.left, target - root.val, temp);
        dfs(root.right, target - root.val, temp);
        temp.remove(temp.size() - 1);
    }
}
